crates/sdk/src/masp.rs

```diff
@@ -583,112 +589,136 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
+                let _ = self.save().await;
+            }
         }
+        Ok(())
     }
 
     /// Fetch the current state of the multi-asset shielded pool into a
     /// ShieldedContext
-    pub async fn fetch<C: Client + Sync>(
+    pub async fn sync<C: Client + Sync>(
         &mut self,
         client: &C,
+        last_block_height: Option<BlockHeight>,
         sks: &[ExtendedSpendingKey],
         fvks: &[ViewingKey],
     ) -> Result<(), Error> {
-        // First determine which of the keys requested to be fetched are new.
-        // Necessary because old transactions will need to be scanned for new
-        // keys.
-        let mut unknown_keys = Vec::new();
+        // add new viewing keys
         for esk in sks {
             let vk = to_viewing_key(esk).vk;
-            if !self.pos_map.contains_key(&vk) {
-                unknown_keys.push(vk);
-            }
+            self.vk_heights.entry(vk).or_default();
         }
         for vk in fvks {
-            if !self.pos_map.contains_key(vk) {
-                unknown_keys.push(*vk);
-            }
-        }
-
-        // If unknown keys are being used, we need to scan older transactions
-        // for any unspent notes
-        let (txs, mut tx_iter);
-        if !unknown_keys.is_empty() {
-            // Load all transactions accepted until this point
-            txs = Self::fetch_shielded_transfers(client, None).await?;
-            tx_iter = txs.iter();
-            // Do this by constructing a shielding context only for unknown keys
-            let mut tx_ctx = Self {
-                utils: self.utils.clone(),
-                ..Default::default()
-            };
-            for vk in unknown_keys {
-                tx_ctx.pos_map.entry(vk).or_insert_with(BTreeSet::new);
-            }
-            // Update this unknown shielded context until it is level with self
-            while tx_ctx.last_indexed != self.last_indexed {
-                if let Some((indexed_tx, (epoch, tx, stx))) = tx_iter.next() {
-                    tx_ctx.scan_tx(*indexed_tx, *epoch, tx, stx)?;
-                } else {
-                    break;
-                }
-            }
-            // Merge the context data originating from the unknown keys into the
-            // current context
-            self.merge(tx_ctx);
-        } else {
-            // Load only transactions accepted from last_txid until this point
-            txs = Self::fetch_shielded_transfers(client, self.last_indexed)
-                .await?;
-            tx_iter = txs.iter();
-        }
-        // Now that we possess the unspent notes corresponding to both old and
-        // new keys up until tx_pos, proceed to scan the new transactions.
-        for (indexed_tx, (epoch, tx, stx)) in &mut tx_iter {
-            self.scan_tx(*indexed_tx, *epoch, tx, stx)?;
+            self.vk_heights.entry(*vk).or_default();
         }
+        // the latest block height which has been added to the witness Merkle
```

Comment:
I'm actually surprised that the integration tests pass. There is bug here that one failed for me on. We need to save the shielded context here to persist the keys.